### PR TITLE
LZ4 decompress: add VBMI optimized copyOverlap32Shuffle

### PR DESCRIPTION
**Pull request by guowangy**
### Changelog category (leave one):
- Performance Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Add VBMI optimized copyOverlap32Shuffle for LZ4 decompress

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

File: src/Compression/LZ4_decompress_faster.cpp [hunk 1]
```diff
@@ -592,6 +652,8 @@ bool decompress(
             success = decompressImpl<8, true>(source, dest, source_size, dest_size);
         if (best_variant == 3)
             success = decompressImpl<32, false>(source, dest, source_size, dest_size);
+        if (best_variant == 4)
```
Comment (alexey-milovidov) [inline src/Compression/LZ4_decompress_faster.cpp:655]: We should activate this method with dynamic dispatching (checking cpuid at runtime).

File: src/Compression/LZ4_decompress_faster.cpp [hunk 2]
```diff
@@ -4,6 +4,7 @@
 #include <iostream>
 #include <Core/Defines.h>
 #include <Common/Stopwatch.h>
+#include <Common/CpuId.h>
```
Comment (alexey-milovidov) [inline src/Compression/LZ4_decompress_faster.cpp:7]: This header is unused. But should be used :)

**Comment by alexey-milovidov** [comment]
Please sign CLA.

**Comment by guowangy** [comment]
@kitaisreal 
> @guowangy did you have any benchmarks when this method provide some improvements ?

This patch can have improvement in `avg_weighted` test:
image

File: src/Compression/LZ4_decompress_faster.cpp (hunk 2 above)
Comment (guowangy) [reply src/Compression/LZ4_decompress_faster.cpp:7]: So keep it here is OK?

File: src/Compression/LZ4_decompress_faster.cpp (hunk 1 above)
Comment (guowangy) [reply src/Compression/LZ4_decompress_faster.cpp:663]: Done. Please help to review.

**Comment by guowangy** [comment]
@alexey-milovidov 

> Please sign CLA.

We have already signed CCLA and just sent it to ClickHouse. I expect it should take some time to take effect.

File: src/Compression/LZ4_decompress_faster.cpp [hunk 3]
```diff
@@ -578,7 +640,13 @@ bool decompress(
     /// Don't run timer if the block is too small.
     if (dest_size >= 32768)
     {
-        size_t best_variant = statistics.select();
+        size_t variant_size = 4;
+#if defined(__AVX512VBMI__) && !defined(MEMORY_SANITIZER)
+        /// best_variant == 4 only valid when AVX512VBMI available
+        if (DB::Cpu::CpuFlagsCache::have_AVX512VBMI)
```
Comment (alexey-milovidov) [inline src/Compression/LZ4_decompress_faster.cpp:646]: Now you have dynamic dispatch but only if the code is compiled with AVX-512 VBMI. But we compile code by default with SSE4.2.

We need to move AVX-512 VBMI code to a separate translation unit, always compile this translation unit with AVX-512 VBMI, and remove `#if defined`.

Comment (guowangy) [reply src/Compression/LZ4_decompress_faster.cpp]: Done, dynamic dispatch with `TargetSpecific`

File: src/Compression/LZ4_decompress_faster.cpp (hunk 2 above)
Comment (guowangy) [reply src/Compression/LZ4_decompress_faster.cpp]: This header removed after using `TargetSpecific`

**Comment by alexey-milovidov** [comment]
Ok, now need to fix the style check and clang-tidy and ARM build.

**Comment by alexey-milovidov** [comment]
The code LGTM.

**Pull request merged**